bergson/callback.py

```diff
@@ -0,0 +1,198 @@
+import os
+
+import numpy as np
+import torch
+import torch.distributed as dist
+from numpy.typing import DTypeLike
+from torch import Tensor
+from transformers.trainer_callback import TrainerCallback, TrainerControl, TrainerState
+from transformers.training_args import TrainingArguments
+
+from bergson import GradientCollector, GradientProcessor
+from bergson.data import create_index
+
+
+class GradientCollectorCallback(TrainerCallback):
+    """Callback that collects gradients from the model during training.
+    Implements the same functionality as `collect_gradients`."""
+
+    def __init__(
+        self,
+        collector: GradientCollector,
+        processor: GradientProcessor,
+        mod_grads: dict[str, Tensor],
+        preconditioners: dict[str, Tensor],
+        grad_sizes: dict[str, int],
+        path: str,
+        dtype: DTypeLike = np.float16,
+        mean_gradients: bool = False,
+        normalize_eval: bool = False,
+    ):
+        """
+        Args:
+            grad_sizes: The sizes of the module gradients
+            processor: The gradient processor
+            mod_grads: The module gradients
+            preconditioners: The gradient preconditioners
+            path: The path to the gradient store
+            dtype: The dtype of the on-disk gradient store
+            mean_gradients: Whether to take the mean of the gradients
+                of the same example across epochs. If `False`, the
+                gradients for each epoch are stored separately.
+            normalize_eval: Whether to normalize the evaluation gradients
+                using the training optimizer normalization.
+        """
+        super().__init__()
+        self.collector = collector
+        self.processor = processor
+        self.mod_grads = mod_grads
+        self.preconditioners = preconditioners
+        self.grad_sizes = grad_sizes
+        self.path = path
+        self.dtype = dtype
+        self.mean_gradients = mean_gradients
+        self.normalize_eval = normalize_eval
+
+    def write_grads(
+        self,
+        grad_buffer: np.memmap,
+        mod_grads: dict[str, Tensor],
+        indices: Tensor,
+    ):
+        for layer_name in mod_grads.keys():
+            torch.cuda.synchronize()
+            grad_buffer[layer_name][indices] = mod_grads[layer_name].numpy()
+        self.mod_grads.clear()
+
+    def on_epoch_begin(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        **kwargs,
+    ):
+        epoch = int(state.epoch or 0)
```

Comment:
Epoch is a float where the decimal part represents progress through the current epoch